datafusion/core/src/datasource/memory.rs

```diff
@@ -143,22 +147,95 @@ impl TableProvider for MemTable {
         _filters: &[Expr],
         _limit: Option<usize>,
     ) -> Result<Arc<dyn ExecutionPlan>> {
+        let batches = &self.batches.read().await;
         Ok(Arc::new(MemoryExec::try_new(
-            &self.batches.clone(),
+            batches,
             self.schema(),
             projection.cloned(),
         )?))
     }
+
+    /// Inserts the execution results of a given [LogicalPlan] into this [MemTable].
+    /// The `LogicalPlan` must have the same schema as this `MemTable`.
+    ///
+    /// # Arguments
+    ///
+    /// * `state` - The [SessionState] containing the context for executing the plan.
+    /// * `input` - The [LogicalPlan] to execute and insert.
+    ///
+    /// # Returns
+    ///
+    /// * A `Result` indicating success or failure.
+    async fn insert_into(&self, state: &SessionState, input: &LogicalPlan) -> Result<()> {
+        // Create a physical plan from the logical plan.
+        let plan = state.create_physical_plan(input).await?;
+
+        // Check that the schema of the plan matches the schema of this table.
+        if !plan.schema().eq(&self.schema) {
+            return Err(DataFusionError::Plan(
+                "Inserting query must have the same schema with the table.".to_string(),
+            ));
+        }
+
+        // Get the number of partitions in the plan and the table.
+        let plan_partition_count = plan.output_partitioning().partition_count();
+        let table_partition_count = self.batches.read().await.len();
+
+        // Adjust the plan as necessary to match the number of partitions in the table.
+        let plan: Arc<dyn ExecutionPlan> =
+            if plan_partition_count == table_partition_count {
+                plan
+            } else if table_partition_count == 1 {
+                // If the table has only one partition, coalesce the partitions in the plan.
+                Arc::new(CoalescePartitionsExec::new(plan))
+            } else {
+                // Otherwise, repartition the plan using a round-robin partitioning scheme.
+                Arc::new(RepartitionExec::try_new(
+                    plan,
+                    Partitioning::RoundRobinBatch(table_partition_count),
+                )?)
+            };
+
+        // Get the task context from the session state.
+        let task_ctx = state.task_ctx();
+
+        // Execute the plan and collect the results into batches.
+        let mut tasks = vec![];
+        for idx in 0..table_partition_count {
```

Comment:
this will try and run all the partitions in parallel, which is probably fine (though maybe we want to limit it to the target_partitions 🤔 )

I don't think this is important to fix now